proposals/0000-a11y-roles-states-improvements.md

```diff
@@ -0,0 +1,106 @@
+---
+title: Improvements to accessibilityRole and accessibilityStates
+author:
+- Marc Mulcahy
+date: 2018-10-31
+---
+
+## Summary
+
+Some brilliant work was begun to add cross platform support for accessibility roles and states to React Native. We propose to add some additional roles and states, as well as extend the implementation on Android and Fire OS.
+
+## Basic example
+
+With the current set of roles, it is not possible to describe a checkbox. These changes would enable this scenario, among other things:
+
+    <View
+      accessibilityRole="checkbox"
+      accessibilityStates={['unchecked']}>
+      <Text>Show Password</Text>
+    </View>
+
+## Motivation
+
+When developers build components in JavaScript, they need the ability to express the role and current state(s) of the component to assistive technologies running on the platform.
+
+## Detailed design
+
+### New Roles
+
+This proposal adds the following roles:
+
+- alert
+- checkbox
+- combobox
+- editabletext
+- menu
+- menubar
+- menuitem
+- progressbar
+- radiobutton
+- radiogroup
+- scrollbar
+- spinbutton
+- tab
+- tablist
+- timer
+- toolbar
+
+### New States
```

Comment:
Web has a lot more states that we'll need to bring over, including `busy`, `expanded`, `hasPopup` etc. I don't know how `on` and `off` map to web ARIA states. Any thoughts?

Reply:
Not sure either ATM. As I mentioned above, maybe we start with ARIA for roles and states, and then add additional states as needed.

Reply:
Yes, these are the the only states we've identified for now. ARIA supports a bunch of states, and it's not always clear how they're supposed to be used. Let's start small, and then add states and support as we find compelling need.

As for on and off, they make sense for things like switches, where the closest analog is probably checked/unchecked, but it doesn't make sense to present something like a light as "checked". This prompts me to think we probably need a "switch" role as well.

As for how to map on/off to web, probably checked/unchecked is the closest...

Reply:
@necolas `hasPopup` is a property in the sense that it designates the possibility of a popup; it is not meant to indicate that a popup is visible or not.